File: crates/vmette/examples/minimal.c
/*
 * Minimal C caller of libvmette. Boots an alpine guest and runs a
 * one-shot command via the C ABI.
 *
 * Build (from repo root):
 *   clang -I crates/vmette/include -L target/release -lvmette \
 *         -Wl,-rpath,@executable_path/../target/release \
 *         -o examples/minimal_c crates/vmette/examples/minimal.c
 *   codesign --sign - --force --entitlements entitlements.plist \
 *            --options=runtime examples/minimal_c
 *
 * Run:
 *   ./examples/minimal_c \
 *       ./assets/vmlinuz-virt ./assets/initramfs-vmette \
 *       ./assets/alpine-rootfs \
 *       'echo "hello from C"; exit 42'
 */

#include <stdio.h>
#include <stdlib.h>
#include "vmette.h"

int main(int argc, char **argv) {
    if (argc < 4) {
        fprintf(stderr, "usage: %s KERNEL INITRAMFS ROOTFS [CMD]\n", argv[0]);
        return 2;
    }
    const char *cmd = (argc >= 5) ? argv[4] : "uname -a; cat /etc/alpine-release; exit 0";

    fprintf(stderr, "vmette version: %s\n", vmette_version());

    vmette_config_t *cfg = vmette_config_new(argv[1], argv[2]);
    if (!cfg) {
        fprintf(stderr, "error: vmette_config_new failed\n");
        return 1;
    }
    vmette_config_set_rootfs_share(cfg, argv[3], /*read_only=*/false);
    vmette_config_set_exec(cfg, cmd);
    vmette_config_set_vsock_port(cfg, 0); /* auto-allocate */

    vmette_run_output_t *out = NULL;
    VmetteStatus rc = vmette_run(cfg, &out);
    /* In the happy path vmette_run never returns — the guest's exit
     * code propagates via the delegate's std::process::exit. We only
     * reach here for snapshot-build/resume paths or hard errors. */
    if (rc != Ok) {
        fprintf(stderr, "vmette_run returned status %d\n", (int)rc);
        vmette_config_free(cfg);
        return 1;
    }
    int code = vmette_run_output_exit_code(out);
    vmette_run_output_free(out);
    vmette_config_free(cfg);
    return code;
}

File: crates/vmette/examples/minimal.rs
// Minimal Rust caller of the vmette library. Boots an alpine guest
// from the repo's existing assets and runs a one-shot command.
//
//   cargo run --release --example minimal -- KERNEL INITRAMFS ROOTFS [CMD]
//
// The binary inherits the parent process's entitlements at run time; the
// example is normally launched from `scripts/run-example.sh` which
// codesigns the cargo-built binary first. Running raw works too if you
// codesign the example binary manually after `cargo build --example`.

use std::env;
use std::process::ExitCode;

use vmette::{Config, RootfsShare, VsockPort};

fn main() -> ExitCode {
    let args: Vec<String> = env::args().skip(1).collect();
    if args.len() < 3 {
        eprintln!("usage: minimal KERNEL INITRAMFS ROOTFS [CMD]");
        return ExitCode::from(2);
    }
    let cmd = args
        .get(3)
        .cloned()
        .unwrap_or_else(|| "uname -a; cat /etc/alpine-release; exit 0".into());

    let mut cfg = Config::new(&args[0], &args[1]);
    cfg.rootfs_share = Some(RootfsShare {
        path: args[2].clone().into(),
        read_only: false,
    });
    cfg.exec_cmd = Some(cmd);
    cfg.vsock_port = VsockPort::Auto;

    match vmette::run(&cfg) {
        // The happy path doesn't actually return — vmette::run exits the
        // process from the VM lifecycle delegate. This arm is for the
        // snapshot-build/resume modes which return synchronously.
        Ok(out) => ExitCode::from(out.exit_code as u8),
        Err(e) => {
            eprintln!("error: {}", e);
            ExitCode::from(1)
        }
    }
}
